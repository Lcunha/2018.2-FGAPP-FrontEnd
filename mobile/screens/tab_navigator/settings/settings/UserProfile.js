import React, { Component } from 'react';
import {
  View,
  StyleSheet,
  Button,
  Image,
  TouchableOpacity,
  Alert,
} from 'react-native';
import { Card, CardItem, Body, Item, Label, Input } from 'native-base';
import jwt_decode from 'jwt-decode'
import UserCard from '../../../components/UserCard'

class UserProfile extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: undefined,
      email: undefined,
      photo: undefined,
      need_logout: false,
    };
  }

  _goBack = async () => {
    const { state } = this.props.navigation;
    var token = state.params ? state.params.token : undefined;

    Alert.alert('Informações atualizadas com sucesso.');
    this.props.navigation.navigate('Settings', { token: token });
  }

  _clickPhoto = async () => {
    const {
      cancelled,
      uri,
    } = await Expo.ImagePicker.launchImageLibraryAsync({ allowsEditing: true, mediaTypes: 'Images', quality: 0.7, base64: true });
    if (!cancelled) {
      this.setState({ photo: uri });
    }
  }
  _logout = async () => {
    const logoutPath = `${process.env.INTEGRA_LOGIN_AUTH}/api/logout/`;
    fetch(logoutPath, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
      }),
    })
    .then((response) => response.json())
    .then((responseJson) => {
      console.log(JSON.stringify(responseJson.detail));
      if (responseJson.detail == 'Successfully logged out.') {
        this.props.navigation.state.params.token = null;
        this.props.navigation.navigate('LoginScreen');
      }
    })
    .catch(err => {
      if (typeof err.text === 'function') {
        err.text().then(errorMessage => {
          this.props.dispatch(displayTheError(errorMessage));
        });
      } else {
        Alert.alert('Erro na conexão.');
        console.log(err);
      }
    });
  }

  _editProfile = async () => {
    const { state } = this.props.navigation;
    var token = state.params ? state.params.token : undefined;

    var user = jwt_decode(token);
    var name = this.state.name;
    var email = this.state.email;
    const uri = this.state.photo;

    const formData = new FormData();
    const updateProfilePath = `${process.env.INTEGRA_LOGIN_AUTH}/api/users/update_profile/`;
    formData.append('user_id', user.user_id);
    if ((name != undefined))
      formData.append('name', name);

    if ((email != undefined)){
      formData.append('email', email);
      this.setState({ need_logout: true });
    }  
    if (uri != undefined) {
      const uriParts = uri.split('.');
      const fileType = uriParts[uriParts.length - 1];
      formData.append('photo', {
        uri: uri,
        name: `photo.${fileType}`,
        type: `application/${fileType}`,
      });
    }

    const options = {
      method: 'POST',
      body: formData,
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'multipart/form-data',
      },
    };

    fetch(updateProfilePath, options)
    .then((response) => {
      return response.json();
    })
    .then((responseJson) => {
      if (responseJson.error != undefined) {
        Alert.alert(responseJson.error);
      }
      else {
        if(this.state.need_logout){
          Alert.alert('Entre no app novamente para aplicar mudanças');
          this._logout();
        }
        else{
          this._goBack();
        }
      }
    })
    .catch(err => {
      if (typeof err.text === 'function') {
        err.text().then(errorMessage => {
          this.props.dispatch(displayTheError(errorMessage));
        });
      } else {
        Alert.alert("Erro na conexão.");
        console.log(err);
      }
    });
  }

  render() {
    const { state } = this.props.navigation;
    const userInfo = state.params ? state.params.userInfo : undefined;
    var name = (this.state.name == undefined) ? userInfo.name : this.state.name;
    var email = (this.state.email == undefined) ? userInfo.email : this.state.email;
    var photo = (this.state.photo == undefined) ? userInfo.photo : this.state.photo;

    return (
      <View style={styles.container}>
        <View style={{ margin: 5 }}>
          <UserCard
            onPress={this._clickPhoto}
            imageSource={{uri: photo}}
            namePlaceholder={name}
            onChangeTextName={(name) => this.setState({name})}
            emailPlaceholder={email}
            onChangeTextEmail={(email) => this.setState({email})}
          />
        </View>
        <View style={{ margin: 5 }}>
          <Button
            color='#BD1C5F'
            onPress={this._editProfile}
            title="Salvar"
          />
        </View>
        <View style={{ margin: 5 }}>
          <Button
            color='#BD1C5F'
            onPress={this._logout}
            title="Sair"
          />
        </View>
      </View>
    );
  }
}
export default UserProfile;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-start',
  },
  backgroundImage: {
    flex: 1,
    position: 'relative'
  },
  image_circle: {
    borderWidth: 1,
    borderColor: 'rgba(0,0,0,0.2)',
    alignItems: 'center',
    justifyContent: 'center',
    width: 100,
    height: 100,
    borderRadius: 100,
  },
  view_circle: {
    alignItems: 'center',
    justifyContent: 'center',
    width: 100,
    height: 100,
    borderRadius: 10,
    borderWidth: 1,
    borderColor: 'rgba(0,0,0,0.2)',
    borderRadius: 100 / 2,
  },
});
