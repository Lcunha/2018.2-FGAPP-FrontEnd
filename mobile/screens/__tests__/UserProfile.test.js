import React from 'react';
import UserProfile from '../tab_navigator/settings/settings/UserProfile';
import Adapter from 'enzyme-adapter-react-16';
import {shallow} from 'enzyme';
import Enzyme from 'enzyme';
import renderer from 'react-test-renderer';
import fetchMock from 'fetch-mock';

Enzyme.configure({adapter: new Adapter()});

describe('Test UserProfile', () => {
    const navigation = {
        state: {
            params: {
                token: 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6InJvZ2VybGVua2VAZ21haWwuY29tIiwidXNlcl9pZCI6MSwib3JpZ19pYXQiOjE1NDE3MTk3NDksImV4cCI6MTU0MTcyMDA0OSwidXNlcm5hbWUiOiJyb2dlcmxlbmtlQGdtYWlsLmNvbSJ9.eCEGRB9yYAkP5iBIybeDsAoWk4HyusPUTX3LBiP0I64',
                userInfo: {
                    name: 'user1',
                    email: 'superuser1@user.com',
                    photo: 'www.notareallink.com/teste1'
                }
            }
        },
        navigate: jest.fn(),
    }

    it('renders correctly', () => {
        const tree = renderer.create(<UserProfile navigation = {navigation}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
})