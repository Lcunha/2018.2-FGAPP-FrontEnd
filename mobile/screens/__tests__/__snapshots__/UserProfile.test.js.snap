// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test UserProfile renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
    }
  }
>
  <View
    style={
      Object {
        "margin": 5,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
            "borderBottomWidth": 0.5,
            "borderColor": "#ccc",
            "borderLeftWidth": 0.5,
            "borderRadius": 2,
            "borderRightWidth": 0.5,
            "borderTopWidth": 0.5,
            "elevation": 3,
            "flexWrap": "nowrap",
            "marginBottom": 5,
            "marginLeft": 2,
            "marginRight": 2,
            "marginTop": 5,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.1,
            "shadowRadius": 1.5,
          },
          Object {
            "height": 150,
            "paddingRight": 10,
          },
        ]
      }
      virtual={undefined}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderRadius": 2,
              "flexDirection": "row",
              "paddingBottom": 10,
              "paddingLeft": 15,
              "paddingRight": 15,
              "paddingTop": 10,
            },
            Object {
              "height": "100%",
            },
          ]
        }
        virtual={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "alignSelf": "stretch",
              "flex": 1,
            }
          }
          virtual={undefined}
        >
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hasTVPreferredFocus={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "borderColor": "rgba(0,0,0,0.2)",
                  "borderRadius": 50,
                  "borderWidth": 1,
                  "height": 100,
                  "justifyContent": "center",
                  "opacity": 1,
                  "width": 100,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View>
                <Image
                  source={
                    Object {
                      "uri": "www.notareallink.com/teste1",
                    }
                  }
                  style={
                    Object {
                      "borderRadius": 100,
                      "height": 100,
                      "position": "absolute",
                      "width": 100,
                    }
                  }
                />
                <Image
                  source={
                    Object {
                      "uri": "https://i.imgur.com/gr7Zvft.png",
                    }
                  }
                  style={
                    Object {
                      "alignItems": "center",
                      "borderColor": "rgba(0,0,0,0.2)",
                      "borderRadius": 100,
                      "borderWidth": 1,
                      "height": 100,
                      "justifyContent": "center",
                      "width": 100,
                    }
                  }
                />
              </View>
            </View>
            <View>
              <View
                accessibilityComponentType={undefined}
                accessibilityLabel={undefined}
                accessibilityTraits={undefined}
                accessible={true}
                collapsable={undefined}
                hasTVPreferredFocus={undefined}
                hitSlop={undefined}
                isTVSelectable={true}
                nativeID={undefined}
                onLayout={undefined}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "borderBottomWidth": 1,
                    "borderColor": "#D9D5DC",
                    "borderLeftWidth": 0,
                    "borderRadius": undefined,
                    "borderRightWidth": 0,
                    "borderTopWidth": 0,
                    "borderWidth": undefined,
                    "flexDirection": null,
                    "marginLeft": 2,
                    "minHeight": 65,
                    "opacity": 1,
                  }
                }
                testID={undefined}
                tvParallaxProperties={undefined}
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "alignSelf": "flex-start",
                        "color": "#575757",
                        "fontSize": 15,
                        "left": null,
                        "paddingRight": 5,
                        "paddingTop": 5,
                        "position": null,
                        "right": null,
                        "top": null,
                      },
                      Object {
                        "fontSize": 12,
                      },
                    ]
                  }
                  virtual={undefined}
                >
                  Nome:
                </Text>
                <TextInput
                  allowFontScaling={true}
                  editable={true}
                  onChangeText={[Function]}
                  placeholder="user1"
                  placeholderTextColor="#575757"
                  style={
                    Array [
                      Object {
                        "alignSelf": "stretch",
                        "color": "#000",
                        "flex": 1,
                        "fontSize": 17,
                        "height": 50,
                        "lineHeight": 18,
                        "paddingLeft": 5,
                        "paddingRight": 5,
                        "top": 1.5,
                        "width": null,
                      },
                      Object {
                        "fontSize": 12,
                      },
                    ]
                  }
                  underlineColorAndroid="rgba(0,0,0,0)"
                  virtual={undefined}
                />
              </View>
              <View
                accessibilityComponentType={undefined}
                accessibilityLabel={undefined}
                accessibilityTraits={undefined}
                accessible={true}
                collapsable={undefined}
                hasTVPreferredFocus={undefined}
                hitSlop={undefined}
                isTVSelectable={true}
                nativeID={undefined}
                onLayout={undefined}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "borderBottomWidth": 1,
                    "borderColor": "#D9D5DC",
                    "borderLeftWidth": 0,
                    "borderRadius": undefined,
                    "borderRightWidth": 0,
                    "borderTopWidth": 0,
                    "borderWidth": undefined,
                    "flexDirection": null,
                    "marginLeft": 2,
                    "minHeight": 65,
                    "opacity": 1,
                  }
                }
                testID={undefined}
                tvParallaxProperties={undefined}
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={
                    Array [
                      Object {
                        "alignSelf": "flex-start",
                        "color": "#575757",
                        "fontSize": 15,
                        "left": null,
                        "paddingRight": 5,
                        "paddingTop": 5,
                        "position": null,
                        "right": null,
                        "top": null,
                      },
                      Object {
                        "fontSize": 12,
                      },
                    ]
                  }
                  virtual={undefined}
                >
                  Email
                </Text>
                <TextInput
                  allowFontScaling={true}
                  editable={true}
                  onChangeText={[Function]}
                  placeholder="superuser1@user.com"
                  placeholderTextColor="#575757"
                  style={
                    Array [
                      Object {
                        "alignSelf": "stretch",
                        "color": "#000",
                        "flex": 1,
                        "fontSize": 17,
                        "height": 50,
                        "lineHeight": 18,
                        "paddingLeft": 5,
                        "paddingRight": 5,
                        "top": 1.5,
                        "width": null,
                      },
                      Object {
                        "fontSize": 12,
                      },
                    ]
                  }
                  underlineColorAndroid="rgba(0,0,0,0)"
                  virtual={undefined}
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "margin": 5,
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#BD1C5F",
              },
            ]
          }
        >
          Salvar
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "margin": 5,
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#BD1C5F",
              },
            ]
          }
        >
          Sair
        </Text>
      </View>
    </View>
  </View>
</View>
`;
